scalafix-core/shared/src/main/scala/scalafix/util/TokenList.scala

```diff
@@ -8,40 +8,29 @@ import scala.meta.tokens.Tokens
 /** Helper to traverse tokens as a doubly linked list.  */
 final class TokenList private (tokens: Tokens) {
   def trailing(token: Token): SeqView[Token, IndexedSeq[Token]] =
-    tokens.view(tok2idx(token), tokens.length).drop(1)
+    tokens.view(tok2idx(token) + 1, tokens.length)
+
   def leading(token: Token): SeqView[Token, IndexedSeq[Token]] =
-    tokens.view(0, tok2idx(token)).drop(1).reverse
+    tokens.view(0, tok2idx(token)).reverse
+
   private[this] val tok2idx = {
     val map = Map.newBuilder[Token, Int]
     var i = 0
     tokens.foreach { tok =>
       map += (tok -> i)
       i += 1
     }
-    map.result()
+    map
+      .result()
+      .withDefault(t =>
+        throw new NoSuchElementException(s"token not found: $t"))
   }
 
-  def find(start: Token)(f: Token => Boolean): Option[Token] = {
-    def loop(curr: Token): Option[Token] = {
-      if (f(curr)) Option(curr)
-      else {
-        val iter = next(curr)
-        if (iter == curr) None // reached EOF
-        else loop(iter)
-      }
-    }
-    loop(next(start))
-  }
+  def find(start: Token)(p: Token => Boolean): Option[Token] =
+    tokens.drop(tok2idx(start)).find(p)
 
-  def slice(from: Token, to: Token): Seq[Token] = {
-    val builder = Seq.newBuilder[Token]
-    var curr = next(from)
-    while (curr.start < to.start) {
-      builder += curr
-      curr = next(curr)
-    }
-    builder.result()
-  }
+  def slice(from: Token, to: Token): Seq[Token] =
+    tokens.view(tok2idx(from), tok2idx(to))
 
```

Comment:
given that the method no longer says that the returned collection is a view, I was going to change the implementation and delegate to `Tokens.slice` instead which should be slightly faster as it is backed by an array and does not involve any copying. But then I realized that `Tokens.length` is buggy, which breaks `isEmpty`:

```
"foo".tokenize.get.slice(1, 0).isEmpty
res4: Boolean = false
```

I'm gonna file an issue or submit a PR to Scalameta.

Reply:
iirc someone reported a similar problem with `Tokens.length` the other day. Length is implemented like this
```scala
  def length: Int = end - start
```

Should it be this instead?

```scala
  def length: Int = math.max(0, end - start)
```

Reply:
The root cause is in `Tokens.slice`: https://github.com/scalameta/scalameta/pull/1207